### PR TITLE
adding DevelopmentSetup.md

### PR DESCRIPTION
**Pull request by fel115**
I created a markdown file to explain the setup of a development environment. This should help developers in the future to set up the environment faster. 

@markus2330 should I create a separate markdown file for troubleshoots?

### CODE REVIEW

File: doc/DevelopmentSetup.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		let text = gettext!("A panic occurred at {}:{}: {}", filename, line, cause);
+		nc.ping(text.clone());
+		eprintln!("{}", text);
+	}));
+``` 
+
+2. If you get an error with "TranslationNotFound("de")", then you have to copy the opensesame.so into your local/de directory
+
+```sh
+cp file/opensesame.so /usr/share/local/de/LC_MESSAGES/
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:70]: Is there no way to tell 

```suggestion
cp file/opensesame.mo /usr/share/local/de/LC_MESSAGES/
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thx for the PR. Great work. Troubleshooting is okay in this file.

File: doc/DevelopmentSetup.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+## 2. Try to build the project (cargo build)
+One importent dependencies is libelektra which can be installed by following [these steps](https://github.com/ElektraInitiative/libelektra/blob/master/doc/INSTALL.md)
+
+## 3. After the build was successfully you have to configure your nextcloud in the files/opensesame.spec file
+Therefor you have to edit the [nextcloud/url], [nextcloud/chat], [nextcloud/chat/licht], [nextcloud/chat/ping] and add [nextcloud/user], [nextcloud/pass]. Instead of the 'required ='-option you have to write 'default ='.
+```sh
+[nextcloud/url]
+description = URL to be used for sending messages.
+default = https://nextcloud.my-server.com/nextcloud
```
[the remark this answers is not in the document]
Comment (fel115) [reply doc/DevelopmentSetup.md:14]: But if I use the required-statment, which is normally in the spec-file, then I get this error: 

`thread 'main' panicked at 'Did not find the key '/sw/libelektra/opensesame/#0/current/nextcloud/url'!', src/config.rs:121:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
   2: opensesame::config::Config::get
             at ./src/config.rs:121:4
   3: opensesame::nextcloud::Nextcloud::new
             at ./src/nextcloud.rs:20:9
   4: opensesame::main
             at ./src/main.rs:187:26
   5: core::ops::function::FnOnce::call_once
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.`

That's the reason why I changed the spec-file. Is there another way to solve this issue?

[the remark this answers is not in the document]
Comment (markus2330) [reply doc/DevelopmentSetup.md:14]: Please create a separate issue with the exact `kdb set` statement you executed.

File: doc/DevelopmentSetup.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# Development Setup 
+For the devolpomnet setup you have to do some steps manually, so this page should help you to get you development environment (devl-env) ready to start.
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:2]:
```suggestion
For the development setup you have to do some steps manually, so this page should help you to get you development environment (devl-env) ready to start.
```

File: doc/DevelopmentSetup.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+The chat-token can be extracted from the chat-url, e.g. `https://nextcloud.my-server.com/nextcloud/index.php/call/<token>`
+
+# Troubleshoot 
+
+1. If you get a panic,with no further information, while running opensesame you have to comment the line 164 out (`panic::set_hook` in main). That's a temporary workarount which is already issued [#6](https://github.com/ElektraInitiative/opensesame/issues/6)
+
+2. If you get an error with "TranslationNotFound("de")", then you have to copy the opensesame.mo into your local/de directory
+
+```sh
+cp file/opensesame.mo /usr/share/local/de/LC_MESSAGES/
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:30]:
```suggestion
cp files/opensesame.mo /usr/share/locale/de/LC_MESSAGES/
```

File: doc/DevelopmentSetup.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/pass "<password>"
+```
+
+The chat-token can be extracted from the chat-url, e.g. `https://nextcloud.my-server.com/nextcloud/index.php/call/<token>`
+
+# Troubleshoot 
+
+1. If you get a panic,with no further information, while running opensesame you have to comment the line 164 out (`panic::set_hook` in main). That's a temporary workarount which is already issued [#6](https://github.com/ElektraInitiative/opensesame/issues/6)
+
+2. If you get an error with "TranslationNotFound("de")", then you have to copy the opensesame.mo into your local/de directory
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:27]:
```suggestion
2. If you get an error with "TranslationNotFound("de")", then you have to copy the opensesame.mo into your locale/de directory
```

File: doc/DevelopmentSetup.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+```sh
+cp file/opensesame.mo /usr/share/local/de/LC_MESSAGES/
+```
+
+3. If you get an IO error in the button.rs file, then you have to [mock the I2CDevice](https://docs.rs/i2cdev/0.5.0/i2cdev/mock/struct.MockI2CDevice.html)
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:33]: How to do that? Do I need to change code in opensesame?

Comment (atmaxinger) [reply doc/DevelopmentSetup.md:33]: Well it's currently running on the Olimex board, and I have an Arduino lying around here too, yes.

Comment (atmaxinger) [reply doc/DevelopmentSetup.md:33]: But sensor-only mode would probably suffice too for my use case

Comment (fel115) [reply doc/DevelopmentSetup.md:33]: I haven't done it until now, but the idea is to create a mock for the I2CDevice so you don't have to have a connection to a i2c-device on your development setup. The error looks like that `Ein unbehebbarer Fehler ist bei src/buttons.rs:132: called `Result::unwrap()` on an `Err` value: Io(Os { code: 13, kind: PermissionDenied, message: "Keine Berechtigung" }) passiert`. 

To answer your question, yes you have to add changes to the opensesame code.

Comment (atmaxinger) [reply doc/DevelopmentSetup.md:33]: Funnily, I have a different error. on a different line

`thread 'main' panicked at 'I2C Communication to Buttons does not work: Nix(ENXIO)', src/buttons.rs:142:14`

File: doc/DevelopmentSetup.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# Development Setup 
+For the development setup you have to do some steps manually, so this page should help you to get you development environment (devl-env) ready to start.
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:2]: What is devl-env? It is only used one more time, so probably not needed to introduce an abbreviation of development environment if this is what you wanted to do.

```suggestion

For the development setup you have to do some steps manually, so this page should help you to get you development environment (devl-env) ready to start.
```

File: doc/DevelopmentSetup.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+## 1. Clone the project on you local devl-env
+
+## 2. Try to build the project (cargo build)
+One importent dependencies is libelektra which can be installed by following [these steps](https://github.com/ElektraInitiative/libelektra/blob/master/doc/INSTALL.md)
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:7]:
```suggestion

One dependency is libelektra which can be installed by following [these steps](https://www.libelektra.org/installation/installation)
```

File: doc/DevelopmentSetup.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+## 3. After the build was successfully you have to configure your nextcloud with `kdb set`
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:9]: First the postinst should be described.


```suggestion

## 3. After the build was successfully you have to configure your nextcloud with `kdb set` 
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Further fixes.

File: doc/DevelopmentSetup.md (hunk 6 above)
Comment (markus2330) [reply doc/DevelopmentSetup.md:33]: > But sensor-only mode would probably suffice too for my use case

So this is best, then you don't need any mocking. Simply install the 
read_analog_input_sketch_aug03a, it should work also without any sensors connected (reading random values if the pins are not connected).

File: doc/DevelopmentSetup.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+The chat-token can be extracted from the chat-url, e.g. `https://nextcloud.my-server.com/nextcloud/index.php/call/<token>`
+
+# Troubleshoot 
+
+1. If you get a panic,with no further information, while running opensesame you have to comment the line 164 out (`panic::set_hook` in main). That's a temporary workarount which is already issued [#6](https://github.com/ElektraInitiative/opensesame/issues/6)
+
+2. If you get an error with "TranslationNotFound("de")", then you have to copy the opensesame.mo into your locale/de directory
+
+```sh
+cp file/opensesame.mo /usr/share/locale/de/LC_MESSAGES/
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:30]:
```suggestion
cp file/opensesame.mo /usr/share/locale/de/LC_MESSAGES/
cp file/opensesame.mo /usr/share/locale/en/LC_MESSAGES/
```

Also add to the English locale as a workaround, as the default Olimex images is in English.

File: doc/DevelopmentSetup.md (hunk 6 above)
Comment (atmaxinger) [reply doc/DevelopmentSetup.md:33]: > read_analog_input_sketch_aug03a

Okay, so if I understand this correctly, the Arduino prints out the measured data from its analog pins onto the USB serial interface. The Arduino is connected to the Olimex board via USB, and opensesame then reads the output of the serial communication.

How do I put opensesame into sensor-only mode?

Alternatively, couldn't I also simulate the I2C buttons using the I2C interface on a Raspberry Pi?

Comment (markus2330) [reply doc/DevelopmentSetup.md:33]: > Arduino prints out the measured data from its analog pins onto the USB serial interface. The Arduino is connected to the Olimex board via USB, and opensesame then reads the output of the serial communication.

Exactly.

> How do I put opensesame into sensor-only mode?

If `sensors/#0/loc` is present, only sensors and env¹ are relevant, then buttons etc. settings are ignored. This is currently an limitation: either sensors or the rest is active.

Btw. there is no spec for sensors/ could you please add this?

¹needs (1) my push to master I just did right before bf4a3865de94d842bb2ef81f1f71fcab78683546 and (2) disabling with environment/device = /dev/null

> Alternatively, couldn't I also simulate the I2C buttons using the I2C interface on a Raspberry Pi?

Of course you could, but this is a lot of useless work (the i2c protocols are quite involved). For testing the mocking is the much better option, as you don't need a complicated setup to make it run.

File: doc/DevelopmentSetup.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# Development Setup 
+For the development setup you have to do some steps manually, so this page should help you to get you development environment (devl-env) ready to start.
+
+## 1. Clone the project on you local devl-env
+
+## 2. Try to build the project (cargo build)
+One importent dependencies is libelektra which can be installed by following [these steps](https://github.com/ElektraInitiative/libelektra/blob/master/doc/INSTALL.md)
+
+## 5. Setup specs and states with Elektra
```
Comment (atmaxinger) [inline doc/DevelopmentSetup.md:9]:
```suggestion
## 3. Setup specs and states with Elektra
```

File: doc/DevelopmentSetup.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+In this step you have to run the `postinst` script, which is located in the debian folder. This script setup the specs, states and mount them into elektra. 
+
+If you want to clear or remove your elektra config you can execute the `postrm` script in the debian folder. 
+
+## 4. After the build and elektra setup was successfully you have to configure your nextcloud with `kdb set` 
+Therefore you have to add [nextcloud/url], [nextcloud/chat], [nextcloud/chat/licht], [nextcloud/chat/ping], [nextcloud/user] and [nextcloud/pass]. This is done with the following statements:
+
+```sh
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/url "https://nextcloud.my-server.com/nextcloud"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/chat "<token>"
```
Comment (fel115) [inline doc/DevelopmentSetup.md:19]: change to `kdb set user:/sw/...`

File: Cargo.toml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "opensesame"
-version = "0.7.6"
+version = "0.7.7"
```
Comment (markus2330) [inline Cargo.toml:3]: This should already be in master, something went wrong with rebase?

File: doc/DevelopmentSetup.md (hunk 12 above)
Comment (markus2330) [reply doc/DevelopmentSetup.md:19]: It is your PR, simply go ahead and push the change :wink:

File: doc/DevelopmentSetup.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
+# Development Setup 
+For the development setup you have to do some steps manually, so this page should help you to get your development environment ready to use.
+
+## 1. Clone the project on you local development environment
+
+## 2. Try to build the project (cargo build)
+One dependency is libelektra which can be installed by following [these steps](https://github.com/ElektraInitiative/libelektra/blob/master/doc/INSTALL.md)
+
+## 3. Setup specs and states with Elektra
+In this step you have to run the `postinst` script, which is located in the debian folder. This script setup the specs, states and mount them into elektra.
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:10]:
```suggestion
In this step you have to run `cd debian && ./postinst`. This script mounts the configuration specification and states into elektra. 
```

File: doc/DevelopmentSetup.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+## 2. Try to build the project (cargo build)
+One dependency is libelektra which can be installed by following [these steps](https://github.com/ElektraInitiative/libelektra/blob/master/doc/INSTALL.md)
+
+## 3. Setup specs and states with Elektra
+In this step you have to run the `postinst` script, which is located in the debian folder. This script setup the specs, states and mount them into elektra. 
+
+If you want to clear or remove your elektra config you can execute the `postrm` script in the debian folder. 
+
+## 4. After the build and elektra setup was successfully you have to configure your nextcloud with `kdb set` 
+Therefore you have to add [nextcloud/url], [nextcloud/chat], [nextcloud/chat/licht], [nextcloud/chat/ping], [nextcloud/user] and [nextcloud/pass]. This is done with the following statements:
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:15]:
```suggestion

As minimum you have to add [nextcloud/url], [nextcloud/chat], [nextcloud/chat/licht], [nextcloud/chat/ping], [nextcloud/user] and [nextcloud/pass]. This is done with the following statements:
```

File: doc/DevelopmentSetup.md [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/chat "<token>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/chat/licht "<token>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/chat/ping "<token>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/user "<user>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/pass "<password>"
+```
+
+The chat-token can be extracted from the chat-url, e.g. `https://nextcloud.my-server.com/nextcloud/index.php/call/<token>`
+
+# Troubleshoot
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:28]: Yes, probably better in a separate document and link there.

File: doc/DevelopmentSetup.md [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/chat/ping "<token>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/user "<user>"
+kdb set system:/sw/libelektra/opensesame/#0/current/nextcloud/pass "<password>"
+```
+
+The chat-token can be extracted from the chat-url, e.g. `https://nextcloud.my-server.com/nextcloud/index.php/call/<token>`
+
+# Troubleshoot 
+
+1. If you get a panic, with no further information, while running opensesame you have to comment the line 164 out (`panic::set_hook` in main). That's a temporary workarount which is already issued [#6](https://github.com/ElektraInitiative/opensesame/issues/6)
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:30]: One line per sentence is easier to review:

```suggestion
1. If you get a panic, with no further information, while running opensesame you have to comment the line 164 out (`panic::set_hook` in main).
That's a temporary workarount which is already issued [#6](https://github.com/ElektraInitiative/opensesame/issues/6).
```

File: doc/DevelopmentSetup.md [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+             at ./src/config.rs:29:5
+   3: opensesame::config::Config::new
+             at ./src/config.rs:20:3
+   4: opensesame::main
+             at ./src/main.rs:141:27
+   5: core::ops::function::FnOnce::call_once
+             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:250:5
+```
+
+The problem can be solved by executing `/var/lib/dpkg/info/opensesame.postinst`.
```
Comment (markus2330) [inline doc/DevelopmentSetup.md:62]: Also add #42 (make sure CWD is the debian folder)

File: doc/Emulating-Olimex.md
Comment (markus2330) [inline doc/Emulating-Olimex.md:1]: This was also in an earlier PR? Please make sure that the [files tab](/https://github.com/ElektraInitiative/opensesame/pull/10/files) contains exactly the wanted changes and nothing else.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Something is still wrong with this PR, please rebase correctly or cherry-pick the relevant changes.

File: doc/DevelopmentSetup.md (hunk 12 above)
Comment (fel115) [reply doc/DevelopmentSetup.md:19]: yeah, I came across this mistake, while I was editing a other PR. I only what to document it for me.

File: Cargo.toml (hunk 13 above)
Comment (fel115) [reply Cargo.toml:3]: Yes, I think so, but I don't know what. I tried to rebase and in the rebase, I wasn't able to select my new changes during the conflict.

File: doc/Emulating-Olimex.md
Comment (fel115) [reply doc/Emulating-Olimex.md:1]: I don´t know what a file tab is.

File: Cargo.toml (hunk 13 above)
Comment (fel115) [reply Cargo.toml:3]: I think now everything should work without conflict.

File: doc/Emulating-Olimex.md
Comment (markus2330) [reply doc/Emulating-Olimex.md:1]: Simply press on the link :wink:

Comment (fel115) [reply doc/Emulating-Olimex.md:1]: Ahh sorry I didn't mention, I get an 404 error by clicking this link

Comment (markus2330) [reply doc/Emulating-Olimex.md:1]: It should've been https://github.com/ElektraInitiative/opensesame/pull/10/files

Comment (fel115) [reply doc/Emulating-Olimex.md:1]: OK now I am understanding what you mean :+1:

**Pull request merged**